src/python_testing/TC_DEM_2_4.py

```diff
@@ -0,0 +1,367 @@
+
+        self.step("9")
+        await self.send_test_event_trigger_user_opt_out_local()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Resumed)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kUserOptOut)
+
+        self.step("9a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("9b")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kOptOut)
+
+        self.step("9c")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+
+        asserts.assert_not_equal(forecast, NullValue)
+        asserts.assert_equal(forecast.forecastUpdateReason,
+                             Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kInternalOptimization)
+
+        self.step("10")
+        await self.send_test_event_trigger_user_opt_out_clear_all()
+
+        self.step("10a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("10b")
+        await self.check_dem_attribute("OptOutState", Clusters.DeviceEnergyManagement.Enums.OptOutStateEnum.kNoOptOut)
+
+        self.step("11")
+        await self.send_pause_request_command(forecast.slots[0].minPauseDuration,
+                                              Clusters.DeviceEnergyManagement.Enums.AdjustmentCauseEnum.kLocalOptimization)
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Paused)
+
+        self.step("11a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kPaused)
+
+        self.step("11b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_equal(forecast.forecastUpdateReason,
+                             Clusters.DeviceEnergyManagement.Enums.ForecastUpdateReasonEnum.kLocalOptimization)
+
+        self.step("12")
+        await self.send_resume_request_command()
+        event_data = events_callback.wait_for_event_report(Clusters.DeviceEnergyManagement.Events.Resumed)
+        asserts.assert_equal(event_data.cause, Clusters.DeviceEnergyManagement.Enums.CauseEnum.kCancelled)
+
+        self.step("12a")
+        await self.check_dem_attribute("ESAState", Clusters.DeviceEnergyManagement.Enums.ESAStateEnum.kOnline)
+
+        self.step("12b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_equal(forecast.isPausable, True)
+        asserts.assert_greater(forecast.slots[0].minPauseDuration, 1)
+        asserts.assert_greater(forecast.slots[0].maxPauseDuration, 1)
+        asserts.assert_equal(forecast.slots[0].slotIsPausable, True)
```

Comment:
Again test plan doesn't ask for this asserts.assert_equal(forecast.slots[0].slotIsPausable, True) (only slot 1). But not a major issue.